README
##########################################################################
#                                                                        #
# (C) Copyright IBM Corporation 2002, 2010 All rights reserved.          #
#                                                                        #
# This program and the accompanying materials are made available under   #
# the terms of the Common Public License v1.0 which accompanies this     #
# distribution, and is also available at http://www.opensource.org       #
# Contributors:                                                          #
#                                                                        #
# William Spurlin - Initial version and framework                        #
#                                                                        #
##########################################################################

 
ClearCase::CtCmd is an interface to IBM Rational ClearCase.  For support, 
contact support@rational.com.  For installation instructions, see
INSTALL.

ClearCase::CtCmd is supported on all platforms that IBM Rational ClearCase
supports.


##########################################################################
#                                                                        #
# Documentation                                                          #
#                                                                        #
##########################################################################

ClearCase::CtCmd is self-documenting through the Perl POD (Plain Old
Documentation) facility.  After installation, "perldoc ClearCase::CtCmd"
will display the self-documentation.  Before installation the command
"perldoc -F CtCmd.pm" may be used from the installation directory.

##########################################################################
#                                                                        #
# Signals                                                                #
#                                                                        #
##########################################################################

ClearCase::CtCmd does not save/restore user signal handlers upon 
entrance/exit.  Users are, of course, able to explictly
restore their signal handlers with the $SIG{<signal>} = <handler ref>
mechanism after each invocation of ClearCase::CtCmd::exec(),
but user signal handlers are modified by the ClearCase libraries
called by ClearCase::CtCmd, except that with ClearCase versions 
greater than 5.0, if the signal handler for signals 1, 2, 3 or 15 
is set to SIG_IGN ( that is, $SIG{[1|2|3|15]} = 'IGNORE' ) 
the signal handler will be saved during and after 
an invocation of ClearCase::CtCmd::exec.

##########################################################################
#                                                                        #
# Limitations                                                            #
#                                                                        #
##########################################################################

# libperl.so

On Solaris and presumably other Unix/Linux, 
/usr/atria/shlib/libperl.<shared object extension> on ClearCase v5.0 
can conflict with the user's perl installation if the user's perl is built 
as a wrapper around  a shared object of the same name, e. g., "libperl.so". 
With IBM Rational ClearCase v6.x and newer, the perl library is name libratlperl.so, removing this conflict.

Workarounds:

A.  Do not build Perl as a shared object.  

    pro:   Simple
    con:   Requires user to rebuild Perl. Since libperl.so is shared, 
           loses advantage of significantly less memory use when
           there are multiple Perl interpreters running. 

B.  Build a shared Perl using some name other than "libperl.<shared object extension>."

    pro:   Simple. Run Configure/modify the config.sh useshrplib,
           libperl and dflt variables, assigning some name other than
           "libperl.so" to libperl. "libperl561.so" does nicely.
    con:   Requires user to rebuild Perl. 

C. Another workaround is for the user (where, for example,  a shared Perl 
    is installed in /opt/perl_5.6.1) to:

    1. From the CtCmd directory, run "perl Makefile.PL; make". 
    2. copy  /usr/atria/shlib/libperl.so to 
        ./blib/arch/auto/ClearCase/CtCmd/liboldperl.so
    3. run the following command:

LD_RUN_PATH="/usr/atria/shlib:/usr/lib:/var/tmp/CtCmd" ld  -G -L/usr/atria/shlib -L/usr/local/lib CtCmd.o -Lblib/arch/auto/ClearCase/CtCmd -R/usr/ucblib  -R/usr/atria/shlib -R/shlib  -R/opt/perl_5.6.1/lib/sun4-solaris/auto/ClearCase/CtCmd  -R/opt/perl/5.6.1/lib/site_perl/auto/ClearCase/CtCmd -Rblib/arch/auto/ClearCase/CtCmd -o blib/arch/auto/ClearCase/CtCmd/CtCmd.so -lzuba  -L/var/tmp/CtCmd  -latriacmdsyn -latriacmd -latriasumcmd -latriasum -latriamsadm -latriamntrpc -latriacm -latriavob -latriaview -latriacm -latriadbrpc -latriatirpc -latriaxdr -latriamvfs -latriatbs -latriaadm -latriasplit -latriacredmap -latriaks -lezrpc -lrpcsvc -latriaccfs -latriasquidad -latriasquidcore -lc -lw -lCrun -loldperl

    (The above will have to be on one line.)

    Now, in lieu of "make test",

    4.  perl -Iblib/lib  -Iblib/arch  t/t1_createArch.t

    5. perl -Iblib/lib  -Iblib/arch  t/t2_tstArch.t

    6. perl -Iblib/lib  -Iblib/arch  t/t3_cleanArch.t

    will run the test suite verbosely and clean up, hopefully with no failures. 

    7.  Copy the contents of  ./blib/arch/auto/ClearCase/CtCmd/ 
    (three files ending in .so ) to 
    /opt/perl_5.6.1/lib/sun4-solaris/auto/ClearCase/CtCmd . 
    Directories will have to be created first. 

    8. Copy blib/lib/ClearCase/CtCmd.pm to  
    /opt/perl_5.6.1/lib/sun4-solaris/ClearCase/CtCmd.pm .  
    Directories will have to be created. 


# AIX

On AIX platforms, with IBM Rational ClearCase v7.x, you must edit 
/usr/opt/perl5/lib[64]/5.8.2/aix-thread-multi/Config.pm and change

lib_ext = ".a"

to

lib_ext = ".so"

before doing "perl Makefile.PL", otherwise perl will look for libraries 
with the ".a" extension.


ChangeLog
##########################################################################
#                                                                        #
# (C) Copyright IBM Corporation 2002, 2012 All rights reserved.          #
#                                                                        #
# This program and the accompanying materials are made available under   #
# the terms of the Common Public License v1.0 which accompanies this     #
# distribution, and is also available at http://www.opensource.org       #
# Contributors:                                                          #
#                                                                        #
# William Spurlin - Creation and updates                                 #
#                                                                        #
# Max Vohlken - Defect fixes                                             #
#                                                                        #
##########################################################################

1.10  Fix for a crash in free() on unix systems using Perl 5.8.6 and later.
      Updated Makefile.PL to deal with new ClearCase library dependencies that 
      were added in ClearCase releases 7.1.2.9 and 8.0.0.5.
1.09  Added a version() subroutine to CtCmd that produces output similar to
      cleartool -verall. This insures that all of the ClearCase core libraries
      will be dynamically loaded when CtCmd is dynamically loaded. Fixed 
      handling of the temporary directory setting in the CC.pm module on 
      Windows. Fixes to the mkfolder and mkstream tests in t1_createArch.t.
      Also this version is included by default in the 8.0 release of 
      ClearCase.
1.08  Added support for AIX 6.x
      This change affects building CtCmd on AIX platforms only.  Changes to
      IBM Rational ClearCase versions 7.0.x and 7.1.x require updating the
      LDDLFLAGS so that CtCmd can be built using /usr/bin/perl on AIX 6.x
1.07  Improved Makefile construction on RHEL5 platform.
      This change affects building CtCmd on RHEL platforms only.  An update
      to /usr/bin/perl requires a change in order to support both RHEL4 and
      RHEL5.  Note that this update also removes -latriatirpc from the 
      library list; MCK is not being updated any more and users can still use 
      CtCmd 1.06 or edit Makefile.PL to restore it.
1.06  Due to changes in release 7 (Baltic) librpcsoc.so is no longer used.
      This change affects only Solaris implementations of CtCmd as 
      librpcsoc.so was only used for Solaris.  This means that release 7
      does not need the aux_zuba.c and the SolarisInterposer code.  Because
      we need backward compatibility that code needs to stay in
      for now.  When releases before release 7 are no longer supported, that
      code and all references to it may be removed.
1.04  Corrected compile warnings generated from aux_zuba.c and proc_table.h.
      License upgraded to CPL v1.0.  On Windows, CtCmd.dll links to ClearCase 
      dlls via symbol names instead of ordinal. So a patch of Clearcase
      will not necessarily require a rebuild of CtCmd. On Solaris, modified 
      aux_zuba.c to fix db_server:RPC failure introduced in ClearCase 2003.06 
      patch 12/13. Some clean up for HP-UX (HP-UX 10.20 no longer supported). 
1.03  Add vob.conf. Expand supported platforms to include Solaris
      2.9, Windows 5.1, HP-UX 11.11.  Modify signal behavior for
      ClearCase > 5.0. Added class method cmdstat() to return error
      status in scalar context.  
1.02  Modify ld in  SolarisInterposer.  Expand Windows $os test. 
      Add sections to README.  Add ChangeLog.
1.01  Fixed outfunc,errfunc.
1.00  Initial release to CPAN.


Makefile.PL
##########################################################################
#                                                                        #
# (C) Copyright IBM Corporation 2001, 2012. All rights reserved.         #
#                                                                        #
# This program and the accompanying materials are made available under   #
# the terms of the Common Public License v1.0 which accompanies this     #
# distribution, and is also available at http://www.opensource.org       #
# Contributors:                                                          #
#                                                                        #
# William Spurlin - Initial version and framework                        #
#                                                                        #
# Xue-Dong Chen - Maintenance and defect fixes                           #
# Max Vohlken - Maintenance and defect fixes                             #
#                                                                        #
##########################################################################

use ExtUtils::MakeMaker;
use File::Basename;

BEGIN {
	# On Windows fix the Config.pm based on where ratlperl is installed.
	if($^O =~ /MSWin32/ || $^O =~ /Windows/) {
		eval "use Win32;";
		my $p = $^X;
		if(basename($p) =~ /^ratlperl(\.exe)?$/) {
			my $i = dirname($p);
			$i = Win32::GetShortPathName($i);
			my $modified = 0;
			my $conf;
			my @conf;
			my @conf_orig;
			for(@INC) {
				if(open(CONF, "$_/Config.pm")) {
					$conf = "$_/Config.pm";
					@conf_orig = <CONF>;
					close(CONF);
					@conf = @conf_orig;
					for(@conf) {
						if(s/\Q\public\ratlperl\5.8.6\nt_i386\E/$i/g) {
							$modified = 1;
						}
					}
				}
			}
			if($conf ne '' and $modified) {
				if(open(CONF, ">$conf")) {
					# If we are able to write to the Config.pm then save
					# a copy of the original as Config.pm.orig.
					if(open(CONF, ">$conf.orig")) {
						print CONF @conf_orig;
						close(CONF);
					}
					print CONF @conf;
					close(CONF);
				} else {
					print "Unable to fix incorrect settings in $conf: $!\n";
				}
			}
		}
	}
}

use Config;

# To override ExtUtils::MM_Unix methods:
# require packages SolarisInterposer(for  MY::subdirs ) 
# and Win32C_O (for MY::c_o).  Including these subs inline
# results in their evaluation at compile time and prohibits platform-selective
# overriding of these methods.
# "use" will not work here as it has an implicit "BEGIN".

my $os=$Config{osname};  #why not $^O ?

my $name=ClearCase::CtCmd;
my $ATRIA_ROOT=$ENV{ATRIA_ROOT}?$ENV{ATRIA_ROOT}:"/opt/rational/clearcase"; 
my $cleartool = $ATRIA_ROOT.'/bin/cleartool';
$cleartool = 'cleartool' if($os =~ /MSWin32/ || $os =~ /Windows/);
my $osver;

die "Cannot find ClearCase 7.0 or higher or cannot find $cleartool" 
unless $ctver = CheckVer('cleartool',
		'7.0',
		$cleartool.' -ver',
		1
		);

my $has_atriacmi = 0;

# libatriacmi was added in 7.1.2.9 and 8.0.0.5.
if(CheckVer(
		'cleartool',
		'7.1.2.9',
		$cleartool.' -ver',
		1
	)
	&& ! CheckVer(
		'cleartool',
		'8.0',
		$cleartool.' -ver',
		1
	)
) {
	$has_atriacmi = 1;
}

if(CheckVer(
		'cleartool',
		'8.0.0.5',
		$cleartool.' -ver',
		1
	)
) {
	$has_atriacmi = 1;
}

if($has_atriacmi) {
	$atriacmi_libs = '-latriacmi -latriajson';
}

my %opts = (
	    'NAME'	=> $name,
	    'VERSION_FROM' => 'CtCmd.pm', # finds $VERSION
	    'LIBS' => "-L. -L$ATRIA_ROOT/shlib -latriacmdsyn -latriacmd -latriasumcmd -latriasum $atriacmi_libs -latriamsadm -latriacm -latriavob -latriaview -latriacm -latriadbrpc -latriaxdr -latriamvfs -latriatbs -latriaadm -latriacredmap -latriaks -lezrpc -lrpcsvc -latriaccfs -latriasquidad -latriasquidcore -latriamsadm -latriamsinfobase -latriamsinfovob"
	    );

$opts{LIBS} .= ' -latriasplit -latriamntrpc' unless($os =~ /MSWin32/ || $os =~ /Windows/);

if ($os =~ /solaris/){
	my($esp_str,$esp_ver,$flag_rd);
	my $default_compiler = $Config{cc};	
    if( $default_compiler =~ /gcc/ || $ENV{CC_CTCMD_GCC}){
	$default_compiler = "gcc";
	$opts{CC} = $default_compiler;
	$opts{LD} = $default_compiler;
	$opts{OPTIMIZE} = '-O';
	$esp_str="gcc version";
	$esp_ver='2.7';
	$flag_rd = ' -v 2>&1';
    }elsif($default_compiler =~ /cc/){
	$esp_str=q(WorkShop Compilers|Sun C);
	$esp_ver='4.2';
	$flag_rd = ' -V 2>&1';
    }else{
	die 'No default compiler';
    }
    if(-e '/usr/lib/libCrun.so.1'){
       if(!-e 'libCrun.so'){
	   if (!symlink('/usr/lib/libCrun.so.1','libCrun.so')){
	       die "/usr/lib/libCrun.so.1: $!";
	   }
       }
    }else{
	die 'No libCrun.so.1 exist on the lib path';
    }
    die 'Cannot find  '.$default_compiler.' -> Sun Workshop Compilers 5.0 or gcc'  
	unless CheckVer($esp_str,
			$esp_ver,
			$default_compiler.$flag_rd
			);
    die "CtCmd requires Solaris 2.7 or higher"
	unless CheckVer("SunOS",
			'5.7',
			'uname -sr'
			);
    $opts{LIBS} ="-L/usr/lib $opts{LIBS} -lratlssl -lc -lw -lCstd -lCrun";
    $opts{DEFINE}	= $opts{DEFINE}.'-DSVR4';
    $opts{INC}	= '-I. ';
    $opts{dynamic_lib} = { OTHERLDFLAGS => "-t -ucmdsyn_proc_table -R$ATRIA_ROOT/shlib"} ;
}elsif ($os =~ /linux/){
    $opts{LD} = $Config{ld};
    die 'Cannot find  '.$Config{cc}  
	unless CheckVer('version',
			'2.7',
			$Config{cc}.' -v 2>&1'
			) ||
	       CheckVer('version egcs-',
			'2.7',$Config{cc}.' -v 2>&1'
		       );
    $opts{DEFINE}= $opts{DEFINE}.'-DATRIA_LINUX';
    $opts{INC}='-I. ';
    $opts{LIBS} ="-L/usr/lib -L/lib $opts{LIBS} -lc -lncurses"; 
    if ($Config{perl} =~ /ratlperl/ || $opts{LD} ne "ld") {
    $opts{dynamic_lib}={ OTHERLDFLAGS => 
			     " -ucmdsyn_proc_table -Wl,-rpath,$ATRIA_ROOT/shlib"
		       };
    } else {
    $opts{dynamic_lib}={ OTHERLDFLAGS => 
                             " -ucmdsyn_proc_table -rpath $ATRIA_ROOT/shlib"
                       };
    }

}elsif  ($os =~ /hpux/){

    $osver = `uname -r`;chomp $osver;
    $osver =~ s/^[A-Z]\.//;print "OSVER $osver\n";
	$mach = `uname -m`; chomp $mach;
    if ($osver == 10.20){
	print "HPUX-10.20\n";
	$opts{LIBS} =  "$opts{LIBS} -lcurses -lM -lezrpc -lrpcsvc -lsec -lstd -lstream -lcl -lc -ldld";
	$opts{DEFINE}= $opts{DEFINE}.' -DATRIA_HPUX10';
	$opts{dynamic_lib}={ OTHERLDFLAGS => ' +b: -E -lCsup'};
    }elsif(11 <= $osver ){
		print "HPUX-11\n";
		if ($Config{perl} =~ /ratlperl/) {
			if($mach eq 'ia64') {
				$opts{LIBS} = "-L/usr/lib/hpux32 $opts{LIBS}";
			} else {
				$opts{LIBS} = "-L/usr/lib $opts{LIBS}";
			}
			$opts{LDFLAGS}=$Config{ldflags};
			$opts{LDFLAGS} =~ s!\-Wl,\+b!-Wl,+b/opt/rational/clearcase/shlib:!g;
			$opts{LDDLFLAGS}=$Config{lddlflags};
			$opts{LDDLFLAGS} =~ s!\-Wl,\+b!-Wl,+b/opt/rational/clearcase/shlib:!g;
		}else{
			#$opts{dynamic_lib}={ OTHERLDFLAGS => ' +b: -E -l:libcl.0  -l:libc.0 '};
		}
		$opts{LIBS} .= " -lc -lstd_v2 -lCsup -lcurses -lrpcsvc -lsec";
    }else{
    }    

}elsif  ($os =~ /MSWin32/ || $os =~ /Windows/){
    die 
	'Cannot find  '
	.$Config{cc}.
	' -> Microsoft (R) 32-bit C/C++ Optimizing Compiler Version 12'  
		unless CheckVer('Version','12',$Config{cc}.' 2>&1' );
    $opts{CCFLAGS} = $Config{ccflags}.' -DATRIA_WIN32_COMMON';
	while($opts{LIBS} =~ /-l(atria[^ ]+)/g) {
		my $lib = $1;
		system($^X, 'autoDef', "win32/$lib.conf") 
		|| system($Config{'ar'}, '/nologo', '/machine:i386', "/def:$lib.def");
	}
    #$opts{LIBS}.=' /defaultlib:WS2_32.LIB';
    $opts{LDDLFLAGS}=$Config{lddlflags};
    $opts{LDDLFLAGS} =~ s/-nodefaultlib//;
    require 'Win32C_O';
}elsif ($os =~ /AIX/i){
    $opts{LIBS} = "-L/usr/lib -L/lib $opts{LIBS} $Config{LIBS}";
    $opts{LDDLFLAGS} = $Config{lddlflags} ." -brtl -bshared ";
    $opts{LDDLFLAGS} =~ s/blibpath:/blibpath:$ATRIA_ROOT\/shlib:/;
}else{print "Unsupported Platform\n"}

if($has_atriacmi) {
	$opts{DEFINE} .= ' -DATRIA_HAS_CMI';
}

WriteMakefile(%opts);

sub CheckVer{
    my $expected_string = shift;
    my $level = shift;
    my $prog_output = shift;
    my $trailing = shift;
    $prog_output = `$prog_output`;
    $prog_output =~ /$expected_string\s*(\d+(\.\d+){1,3})/i;
    $found_level = $1;
    @minversion = split('\.',$level);
#
# The output of the cleartool -ver command has changed.
# An earlier version looks like 2003.06.00 while a later version
# looks like 7.0.0.0   We need to be able to handle both styles. 
# The trailing flag means that the version trails the year so we
# need to remove the year.  That is no longer always true so we
# need to test to see if the year is first, if it is not the we
# have to turn off the trailing flag. 
    if($trailing){
    if($found_level !~ /200/) {
    $trailing = 0;
    }
    }

    if($trailing){
	$found_level =~ s/\d+\.(\d+\.\d+)/$1/;
    }
    @foundversion = split('\.',$found_level);
    my $ans = Found;
    $ans = 'Not Found' unless $1;
    print "Looking for $expected_string\t$level\t$ans\t$found_level\n";

    for ($i = 0; $i < @minversion; $i++)
        {
           if ($foundversion[$i] > $minversion[$i]){
                return  $found_level;
           }else{   
                if ($foundversion[$i] < $minversion[$i]){ 
                   return 0;
                }
           }
     }
     return  $found_level;
}


